KEP 4960: Container Stop Signals

sreeram-venkitesh opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
/kind api-change
/sig node

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR includes the implementation for the alpha realease of https://github.com/kubernetes/enhancements/issues/4960. We're adding a new Container Lifecycle, StopSignal, with which users can configure custom stop signals for their containers. This involves changes to the CRI API as well.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/enhancements/issues/4960

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added support for configuring custom stop signals with a new StopSignal container lifecycle
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- KEP: https://github.com/kubernetes/enhancements/issues/4960
```

mrunalp commented:
@sreeram-venkitesh Could you move this from draft? Reviewers won't typically look at it if it is in draft.

sreeram-venkitesh commented:
There are a bunch of tests which will fail now. I had kept the PR in draft until the tests were fixed. I will push the test fixes later today.

aaron-prindle commented:
/remove-sig api-machinery

mrunalp commented on pkg/apis/core/types.go:
```diff
@@ -2673,6 +2673,47 @@ type GRPCAction struct {
+type Signal string
+
+const (
+	Sigabrt   Signal = "SIGABRT"
```
We could consider keeping these names as full caps.

thockin replied:
+1  It's unusual but the convention of all caps is very strong for signals.  Even Go does it:

```
$ go doc -all unix | grep '^[   ]*SIG'
	SIGABRT = syscall.Signal(0x6)
	SIGALRM = syscall.Signal(0xe)
	SIGFPE  = syscall.Signal(0x8)
	SIGHUP  = syscall.Signal(0x1)
	SIGILL  = syscall.Signal(0x4)
	SIGINT  = syscall.Signal(0x2)
	SIGIOT  = syscall.Signal(0x6)
	SIGKILL = syscall.Signal(0x9)
	SIGPIPE = syscall.Signal(0xd)
	SIGQUIT = syscall.Signal(0x3)
	SIGSEGV = syscall.Signal(0xb)
	SIGTERM = syscall.Signal(0xf)
	SIGTRAP = syscall.Signal(0x5)
	SIGBUS    = syscall.Signal(0x7)
	SIGCHLD   = syscall.Signal(0x11)
	SIGCLD    = syscall.Signal(0x11)
	SIGCONT   = syscall.Signal(0x12)
	SIGIO     = syscall.Signal(0x1d)
	SIGPOLL   = syscall.Signal(0x1d)
	SIGPROF   = syscall.Signal(0x1b)
	SIGPWR    = syscall.Signal(0x1e)
	SIGSTKFLT = syscall.Signal(0x10)
	SIGSTOP   = syscall.Signal(0x13)
	SIGSYS    = syscall.Signal(0x1f)
	SIGTSTP   = syscall.Signal(0x14)
	SIGTTIN   = syscall.Signal(0x15)
	SIGTTOU   = syscall.Signal(0x16)
	SIGURG    = syscall.Signal(0x17)
	SIGUSR1   = syscall.Signal(0xa)
	SIGUSR2   = syscall.Signal(0xc)
	SIGVTALRM = syscall.Signal(0x1a)
	SIGWINCH  = syscall.Signal(0x1c)
	SIGXCPU   = syscall.Signal(0x18)
	SIGXFSZ   = syscall.Signal(0x19)
	SIG_BLOCK   = 0x0
	SIG_UNBLOCK = 0x1
	SIG_SETMASK = 0x2

```

mrunalp commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3352,14 +3352,116 @@ func validateHandler(handler commonHandler, gracePeriod *int64, fldPath *field.P
+func validateStopSignal(stopSignal *core.Signal, fldPath *field.Path, os *core.PodOS) field.ErrorList {
+	allErrors := field.ErrorList{}
+
+	if os == nil {
```
I think we can drop this and assume linux as default.
cc: @thockin

liggitt replied:
I really want to start requiring explicit OS indication for use of OS-specific fields

thockin replied:
So we should validate this as "if you didn't say Linux, you don't get to use Linux-isms?"  I think I am OK with that.

liggitt replied:
to use new OS-specific fields, yeah, I'd lean that way

sreeram-venkitesh replied:
I've added the note in https://github.com/kubernetes/kubernetes/pull/130556/commits/caeeb362b68e3db945bc68c68d45829636ab7746. PTAL!

mrunalp commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -256,6 +256,7 @@ func testPodDefaults(t *testing.T, featuresEnabled bool) {
 		".Spec.Containers[0].Lifecycle.PostStart.HTTPGet.Scheme":         `"HTTP"`,
 		".Spec.Containers[0].Lifecycle.PreStop.HTTPGet.Path":             `"/"`,
 		".Spec.Containers[0].Lifecycle.PreStop.HTTPGet.Scheme":           `"HTTP"`,
+		".Spec.Containers[0].Lifecycle.StopSignal":                       `"SIGTERM"`,
```
Agree!

sreeram-venkitesh replied:
Yep, this failing test was resolved once I used `ContainerLifecycleStaus` instead of the whole `Lifecycle` struct as you mentioned in https://github.com/kubernetes/kubernetes/pull/130556#discussion_r1994182456. Removed the defaults in https://github.com/kubernetes/kubernetes/pull/130556/commits/b101b2489c80bb8f23bbee51abe4db3b4297fe00.

sreeram-venkitesh commented on pkg/apis/core/types.go:
```diff
@@ -2831,6 +2875,10 @@ type ContainerStatus struct {
+	// Lifecycle is used to show the container's effective stop signal in the Status
+	// +featureGate=ContainerStopSignals
+	// +optional
+	Lifecycle Lifecycle
```
Done in https://github.com/kubernetes/kubernetes/pull/130556/commits/4fa161fb775bb535d2bceeed2d7c62eed7791e54. This change also fixes some failing tests.

sreeram-venkitesh commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3154,6 +3198,10 @@ type ContainerStatus struct {
 	// +listType=map
 	// +listMapKey=name
 	AllocatedResourcesStatus []ResourceStatus `json:"allocatedResourcesStatus,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,14,rep,name=allocatedResourcesStatus"`
+	// Lifecycle is used to show the container's effective stop signal in the Status
```
Updated [here](https://github.com/kubernetes/kubernetes/pull/130556/commits/4fa161fb775bb535d2bceeed2d7c62eed7791e54#diff-5ffbebb40a29f8e491f4ff1e07c618eb3edcc0d5f00a1a71445272d6767d7883).

sreeram-venkitesh commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3352,14 +3352,116 @@ func validateHandler(handler commonHandler, gracePeriod *int64, fldPath *field.P
 
-func validateLifecycle(lifecycle *core.Lifecycle, gracePeriod *int64, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+var supportedStopSignalsLinux = sets.New(
+	core.Sigabrt,
```
Done in https://github.com/kubernetes/kubernetes/pull/130556/commits/f4fb222f808ea549ec98edf406b253502db72af3.

sreeram-venkitesh commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3352,14 +3352,116 @@ func validateHandler(handler commonHandler, gracePeriod *int64, fldPath *field.P
+	if os == nil {
+		allErrors = append(allErrors, field.Forbidden(fldPath, "may not be set for containers with empty .spec.os.name"))
+	} else if os.Name == core.Windows {
+		switch *stopSignal {
```
Done in https://github.com/kubernetes/kubernetes/pull/130556/commits/befc39ce56dcb49362a8d20dd680370d97c53ba9.

sreeram-venkitesh commented on pkg/api/pod/util.go:
```diff
@@ -678,6 +678,58 @@ func dropDisabledFields(
+
+	adjustLifecycle := func(lifecycle *api.Lifecycle) {
+		if lifecycle.StopSignal != nil {
+			lifecycle.StopSignal = nil
```
Done in https://github.com/kubernetes/kubernetes/pull/130556/commits/befc39ce56dcb49362a8d20dd680370d97c53ba9.

thockin replied:
Sorry, what I meant was something like:

```
wipeLifecycle := func(ctr *api.Container) {
    if ctr.Lifecycle == nil {
        return
    }
    if ctr.Lifecycle.StopSignal != nil {
        ctr.Lifecycle.StopSignal = nil
        if *ctr.Lifecycle == api.Lifecycle{} {
            // stopSignal was the only thing in lifecycle
            ctr.Lifecycle = nil
        }
    }
}
```

sreeram-venkitesh replied:
I've made this change in https://github.com/kubernetes/kubernetes/pull/130556/commits/a857f239816bf42ecbf58033ae7d1076be85de34.

thockin commented on pkg/api/pod/util.go:
```diff
@@ -678,6 +678,58 @@ func dropDisabledFields(
+		}
+	}
+
+	for i := range podSpec.Containers {
```
I think you just reinvented `VisitContainers()` - can you use that instead?

sreeram-venkitesh replied:
Done in https://github.com/kubernetes/kubernetes/pull/130556/commits/a857f239816bf42ecbf58033ae7d1076be85de34. This was how it was done in the `dropPodLifecycleSleepAction`, we can probably refactor that function too.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2831,6 +2877,20 @@ type ContainerStatus struct {
+	Lifecycle ContainerLifecycleStatus
+}
+
+// ContainerLifecycleStatus represents the effective stop signal used by the container runtime.
```
Picking at words:

```
// ContainerLifecycleStatus reports information about the lifecycle management of one container.
type ContainerLifecycleStatus struct {
    // StopSignal is the signal which will be sent to this container when it is stopped.  This might be
    // the stop signal specified by the user, the signal specified in the container image, or the default
    // stop signal of the container runtime on this node.
```
```

sreeram-venkitesh replied:
Done in https://github.com/kubernetes/kubernetes/pull/130556/commits/6041eba787c31598f519ab4de14edd2883500431.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3342,14 +3342,48 @@ func validateHandler(handler commonHandler, gracePeriod *int64, fldPath *field.P
+	allErrors := field.ErrorList{}
+
+	if os == nil {
+		allErrors = append(allErrors, field.Forbidden(fldPath, "may not be set for containers with empty .spec.os.name"))
```
```
... "may not be set for containers with empty `spec.os.name`"
```

sreeram-venkitesh replied:
Done in https://github.com/kubernetes/kubernetes/pull/130556/commits/6041eba787c31598f519ab4de14edd2883500431.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3352,14 +3352,116 @@ func validateHandler(handler commonHandler, gracePeriod *int64, fldPath *field.P
+		default:
+			allErrors = append(allErrors, field.NotSupported(fldPath, stopSignal, sets.List(supportedStopSignalsWindows)))
+		}
+	} else if os.Name == core.Linux {
```
I still do not see testcases for this

sreeram-venkitesh replied:
Sorry, I hadn't pushed last day since I was still working on it. I have pushed unit tests in https://github.com/kubernetes/kubernetes/pull/130556/commits/77e467d4e549f5f8d1b557c3bd07fe7171a64774.

ffromani commented:
/triage accepted
/priority important-soon

we agreed to have this in 1.33

ffromani commented:
/cc

sreeram-venkitesh commented:
Also, I need help with fixing the linter hints errors. I had added lifecycle in the expected Pod status here to fix some failing tests. The error says:
```
ERROR: pkg/util/pod/pod_test.go:87:11: S1039: unnecessary use of fmt.Sprint (gosimple)
ERROR: 			[]byte(fmt.Sprint(`{"metadata":{"uid":"myuid"},"status":{"initContainerStatuses":[{"image":"","imageID":"","lastState":{},"lifecycle":{},"name":"init-container","ready":true,"restartCount":0,"state":{}}]}}`)),
ERROR: 			       ^
Please review the above warnings. You can test via "./hack/verify-golangci-lint.sh -r 381ccf0f4c723f7164d4363d2bcd76f4e34e5441 -n"
If the above warnings do not make sense, you can exempt this warning with a comment
 (if your reviewer is okay with it).
```
I had tried changing `fmt.Sprintf` to `fmt.Sprint` but its showing the same error now for `fmt.Sprint`. How can I fix/disable this? I tried searching GitHub and Slack for similar threads but couldn't find anything super useful. Thanks in advance!

thockin commented:
Can you squash it down to a meaningful set of commits befor merge, please?

thockin commented:
> I had tried changing fmt.Sprintf to fmt.Sprint but its showing the same error now for fmt.Sprint. How can I fix/disable this? 

`x = fmt.Sprint("foo")` is exactly the same as `x = "foo"` - it's just a string.

thockin commented:
API parts LGTM.  Let me know when it is approved for logic and I will final approve

sreeram-venkitesh commented:
Thanks! I'll squash the commits before merge.

> > I had tried changing fmt.Sprintf to fmt.Sprint but its showing the same error now for fmt.Sprint. How can I fix/disable this?

>  x = fmt.Sprint("foo") is exactly the same as x = "foo" - it's just a string.

I can make this change if that fixes the errors, but cc-ing @liggitt here since these were added in https://github.com/kubernetes/kubernetes/pull/86320. Please check https://github.com/kubernetes/kubernetes/pull/130556#issuecomment-2734397649. Thanks in advance!

thockin commented:
Probably was added before lint was enforced and probably was printf which lost its last expansion at some point, and now lint is saying "this does nothing"

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1359,6 +1366,13 @@ message ContainerStatus {
     ContainerUser user = 18;
+
+    // Returns the stop signal used by the container runtime to terminate the container 
+    ContainerLifecycleStatus lifecycle = 19;
```
We could directly add the stop signal. I am not seeing a strong need to nest under Lifecycle message here.

sreeram-venkitesh commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1186,6 +1186,13 @@ message ContainerConfig {
+}
+
+message Lifecycle {
+  string stop_signal = 1;
```
@mikebrow [had said that](https://kubernetes.slack.com/archives/CJUQN3E4T/p1741888560417639?thread_ts=1741696514.280119&cid=CJUQN3E4T) he prefers it being a string over an enum since nil/unset values can mean the runtime can fallback to the signal in the image/runtime default.

I had tried implementing an enum here, but when the stopsignal in the container lifecycle was empty, the signal was defaulting to the zero value of the enum when it gets passed to the runtime. We'd need to add a placeholder signal like SIGNONE in the CRI enum alone, so that empty lifecycle would be represented with SIGNONE and the runtime can unpack the signal defined in the image if it receives SIGNONE in the ContainerConfig. [Related discussion](https://kubernetes.slack.com/archives/C0BP8PW9G/p1741626284578169).

I'm all in for an enum, if I can go ahead with this logic. Please let me know your thoughts!

mikebrow replied:
additionally the valid values are per platform/arch or we are going to create a k8s set of values and have to map them at some point before they get used.. Thus was having difficulty seeing how we could/should do just one set of enums at the cri layer..

sreeram-venkitesh replied:
Currently we have a list of supported signals for Linux and Windows at the Kubernets validation level with which we are cross validating the signal against the Pod's `spec.os.name`. If we want to do something similar for different architectures within Linux, we'd need to have multiple enums at the CRI layer, have I understood this correctly?
 
https://github.com/kubernetes/kubernetes/blob/927a0363e79ed7c5a03ccbb1a0b0e8435b68e938/pkg/apis/core/validation/validation.go#L3345-L3356

mrunalp replied:
What if we have an enum with all these signals prefixed with SIGNAL_ for e.g. `SIGNAL_SIGABRT = 1, SIGNAL_SIGALRM = 2 and so on and also add `SIGNAL_RUNTIME_DEFAULT / SIGNAL_NONE = 0`? 

That same list could work for both windows/linux. 

@mikebrow wdyt?

mikebrow replied:
- that would cover most of the posix compliant systems linux, openbsd, freebsd (maybe darwin some day)...
- to fill out posix systems we would also need SIGWINCH (terminal resize window change), SIGXCPU (CPU time limit exceeded), SIGXFSZ (file size limit exceeded),.. and SIGRTMIN..SIGRTMAX and the values between normally referred as "RTMIN+1" or sigrtmin+1... repeat up to +15... then count down from "RTMAX-14" to "RTMAX-1"
- Some of these signals are not really appropriate for use as stop signals.. 

for windows I'd be concerned, architecturally, that we may be filtering down the docker/moby list: https://github.com/moby/sys/blob/signal/v0.7.1/signal/signal_windows.go  Though for "basic" support that would be fine because it is the default signal set for windows that we are using.

Also I'm concerned about pushing ever more platform information into the CRI client (kubelet) when that may be configured deep in a VM wrapped WASI plugin :-)

mikebrow replied:
the 0 value is default.. on a new field can work.. since the old CRI servers in the field will not be using the new field with the gate on or off and the the values here will be iota's that will get mapped to their real values anyways.. 

Was just trying to save some code iteration by using the pre-existing map of strings pattern. But yeah this works and saves a couple bytes :-)

thockin commented:
reminder: code freeze in about 30 hours

thockin commented:
Are we on track to finish this in the next 7 hours?

sreeram-venkitesh commented:
@thockin I'm working on pushing the changes suggested for the CRI API and the replacing ContainerLifecycleStatus with StopSignal in the status.

haircommander commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -663,6 +665,21 @@ func toKubeContainerStatus(status *runtimeapi.ContainerStatus, runtimeName strin
+		}
+	} else {
+		cStatusLifecycle = &v1.Lifecycle{}
+	}
```
nil seems a bit more correct to me but I think this is okay too

haircommander commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -663,6 +665,21 @@ func toKubeContainerStatus(status *runtimeapi.ContainerStatus, runtimeName strin
+			cStatusLifecycle = &v1.ContainerLifecycleStatus{}
+		}
+	} else {
+		cStatusLifecycle = &v1.ContainerLifecycleStatus{}
```
this case feels like it should be left nil

mrunalp commented:
@sreeram-venkitesh As you work on addressing the review comments, could you also clean up the commits? Thanks!

sreeram-venkitesh commented:
Folks, I've addressed all the comments and have squashed my commits. I'm still thinking to request for a code freeze exception so that I can review and test everything once again tomorrow morning. I also want to refactor [these helper functions](https://github.com/kubernetes/kubernetes/pull/130556/commits/e17bd5588910003f25c23e3db414d7b3f8ee5ce8#diff-8af985f57262855ea5040fd763c1e5606085928c4efece7d30d31352e893218a). I wouldn't need more than a day, but I'm planning to ask until Monday (March 24th) to be sure. I had discussed this with @haircommander.

CC: @thockin @mikebrow @mrunalp Thanks a lot for the reviews and all the help so far!

thockin commented:
I'm standing by to approve the kube-API parts once the rest is approved

HirazawaUi commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -668,6 +670,17 @@ func toKubeContainerStatus(status *runtimeapi.ContainerStatus, runtimeName strin
 		cStatusUser = toKubeContainerUser(status.User)
 	}
 
+	var cStatusStopSignal *v1.Signal
+	if utilfeature.DefaultFeatureGate.Enabled(features.ContainerStopSignals) {
+		if signal := status.GetStopSignal().String(); signal != "" {
+			cStatusStopSignal = fromCRIStopSignal(status.GetStopSignal())
+		} else {
+			cStatusStopSignal = nil
+		}
+	} else {
+		cStatusStopSignal = nil
+	}
+
```
```suggestion
var cStatusStopSignal *v1.Signal
if utilfeature.DefaultFeatureGate.Enabled(features.ContainerStopSignals) {
    signal := status.GetStopSignal().String()
    if signal != "" {
        cStatusStopSignal = fromCRIStopSignal(status.GetStopSignal())
    }
}
```
This will make the code look simpler.

HirazawaUi commented:
You also need to locally execute `hack/update-openapi-spec.sh` and fix the unit test errors.

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1186,6 +1186,83 @@ message ContainerConfig {
+}
+
+enum Signal {
+    RUNTIME_DEFAULT   = 0;
```
Alignment is off.

sreeram-venkitesh replied:
Please check https://github.com/kubernetes/kubernetes/pull/130556/commits/c8a8b89ef6643dee782d7fedf09cfef5ec41568b.

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1186,6 +1186,83 @@ message ContainerConfig {
+	SIGWINCH          = 32;
+	SIGXCPU           = 33;
+	SIGXFSZ           = 34;
+    SIGRTMIN          = 35;
```
Alignment is off for the next block of lines.

sreeram-venkitesh commented:
Quick update: 
- Code freeze exception has been approved until 9:00 PDT Monday.
- I've added an e2e test and have been testing the feature more locally - flipping feature gate off and on again etc.
- Still working on cleaning up pkg/kubelet/kuberuntime/helpers.go, which I should be able to finish today.

CC: @dipesh-rawat

sreeram-venkitesh commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1186,6 +1186,83 @@ message ContainerConfig {
     repeated CDIDevice CDI_devices = 17;
+
+    // Lifecycle includes the custom stop signal for the container
+    Lifecycle lifecycle = 18;
```
I'd like to keep StopSignal nested under Lifecycle in the Kubernetes API. Can I change just the CRI API to take `Containers[i].Lifecycle.StopSignal` and pass it to `ContainerConfig.StopSignal`? In the similar comment for StopSignal in ContainerStatus above, I had updated both the K8s API and the CRI API to have `Status.StopSignal` instead of `Status.Lifecycle.StopSignal`, but this I'm more okay with since Status didn't already have a Lifecycle. 

Please let me know what you think!

cc: @thockin

thockin replied:
In k8s API it makes sense under lifecycle.  I have less visibility into CRI and so do not express an opinion on that.

sreeram-venkitesh commented:
The test failures seems to be unrelated, will try retesting onve everything finishes.

sreeram-venkitesh commented:
/retest

sreeram-venkitesh commented:
I've pushed the refactoring I wanted to do, I'm waiting for comments on https://github.com/kubernetes/kubernetes/pull/130556#discussion_r2008623520 now.

sreeram-venkitesh commented:
/retest

sreeram-venkitesh commented:
/retest

toVersus commented on test/e2e/common/node/lifecycle_hook.go:
```diff
@@ -661,3 +662,43 @@ var _ = SIGDescribe(feature.PodLifecycleSleepActionAllowZero, func() {
 	})
 })
+
+var _ = SIGDescribe(feature.ContainerStopSignals, func() {
```
Since the container runtime implementation for the CRI API change hasn’t been done yet, does that mean this release will only test whether StopSignal can be used at the API level here? It might not be appropriate to bring this up with the exception request deadline approaching, but I was concerned because it doesn’t seem to follow the [Kubernetes feature development and container runtimes policy](https://www.kubernetes.dev/docs/code/cri-api-dev-policies/#same-maturity-level-for-alpha).

sreeram-venkitesh replied:
Thanks for bringing this up! This only tests StopSignal at the API level right now. I can remove the test if it makes more sense to have this once the container runtime also supports StopSignal.

thockin commented:
Assuming the kube-API parts are unchanged, I am waiting for CRI to be LGTM from... @mrunalp I think? 

This issue is not assigned to anyone.

sreeram-venkitesh commented:
@thockin: @mrunalp's comment https://github.com/kubernetes/kubernetes/pull/130556#discussion_r2008623520 needs to be resolved. If what I have posted is acceptable, we can make the change in CRI API and leave kube-API unchanged.

thockin commented:
Sorry, somehow comments were not sent.

thockin commented:
I'm going to approve for API now, I will be unavailable for a while and would not want this to miss because of me.

/approve

sreeram-venkitesh committed:
API changes

sreeram-venkitesh committed:
Validation

sreeram-venkitesh committed:
CRI API changes

sreeram-venkitesh committed:
Implementation

sreeram-venkitesh committed:
Added e2e test and fixed existing pod test

sreeram-venkitesh committed:
Generated files

sreeram-venkitesh committed:
Generated UPDATE_COMPATIBILITY_FIXTURE_DATA

haircommander commented:
/lgtm

great work @sreeram-venkitesh !

haircommander commented:
/milestone v1.33

(I don't think I'm allowed to do this but just checking)

sreeram-venkitesh commented:
The test failure seems to be unrelated. Will try retesting once all the checks are completed.

sreeram-venkitesh commented:
/retest

sreeram-venkitesh commented:
/retest
